Add ksmConfiguration to enable ksm in the nodes

fossedihelm opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Follow the instructions for writing a release note from k8s: https://git.k8s.io/community/contributors/guide/release-notes.md
-->

**What this PR does / why we need it**:
With `ksmConfiguration` cluster admins are able to define a set of nodes in which ksm will be enabled.
The nodes will be filtered by label with the LabelSelector. If a node matches the labels and has the ksm available, the ksm in that node will be enabled. Alongside the enabling of the ksm, an annotation will be added to the node.
This annotation marks nodes where the ksm has been enabled by the virt-handler, allowing kv to disable it in case a subsequent change in the selector will cause a node to no longer be included.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
Cluster admins can enable ksm in a set of nodes via kv configuration
```

/cc @xpivarc @acardace @vladikr @Barakmor1

fossedihelm commented:
/retest-required

alicefr commented:
@fossedihelm I have a question. Why don't we use a label instead of an annotation? I wondering because if we use a label then we could use it to schedule the VM and filtering the nodes with ksm.

fossedihelm commented:
> @fossedihelm I have a question. Why don't we use a label instead of an annotation? I wondering because if we use a label then we could use it to schedule the VM and filtering the nodes with ksm.

@alicefr A label for those nodes has been introduced in https://github.com/kubevirt/kubevirt/pull/9691 :)
This annotation is used to track which nodes has been modified by the virt-handler, so that the ksm will be disabled in case of changes in the selector terms. For example when the ksm is first enabled in all the nodes, but then a filter is introduced: in this case, in the nodes in which ksm was first enabled, it will be disabled. Hope this helps

alicefr commented:
@fossedihelm sorry, I might missing an essential point but I fail to understand why we need 2 ways to mark the node. Isn't a single label sufficient? Or is it to mark which nodes have already been updated?

acardace commented on tests/infra_test.go:
```diff
@@ -1690,6 +1690,112 @@ var _ = Describe("[Serial][sig-compute]Infrastructure", Serial, decorators.SigCo
+		BeforeEach(func() {
+			nodesWithKSM := getNodesWithKSMAvailable(virtClient)
+			if len(nodesWithKSM) == 0 {
+				Skip("Skip testing with ksm, because there are no nodes with ksm available")
```
Shouldn't we use Ginkgo labels/decorators instead of blindly skipping?

fossedihelm replied:
I had thought about it but couldn't think of a way to activate that label. Now that you've pointed it out to me again, it's quite simple actually. Thanks

fossedihelm replied:
Done

acardace commented:
> @fossedihelm sorry, I might missing an essential point but I fail to understand why we need 2 ways to mark the node. Isn't a single label sufficient? Or is it to mark which nodes have already been updated?

Hi Alice, 
I'll try to answer, the annotation is an internal record to keep track of which nodes are currently managed by virt-handler (in regard to KSM), and the label is to be used by VMs for scheduling purposes.

The label itself could be present also on nodes where KSM was not enabled and thus managed by KubeVirt.

alicefr commented:
@acardace  thanks :) it is clearer now

fossedihelm commented on staging/src/kubevirt.io/api/core/v1/types.go:
```diff
@@ -2573,10 +2578,19 @@ type NodeMediatedDeviceTypesConfig struct {
+	// Selector is a selector that filters in which nodes the KSM will be enabled.
+	// Empty Selector will enable ksm for every node.
+	// +optional
+	Selector *metav1.LabelSelector `json:"selector,omitempty"`
```
IMHO `metav1.LabelSelector` allows to define more complex matching rules (i.e. the `In` operator will allow to do a sort of OR operation). Also, it is totally aligned with the k8s syntax

acardace replied:
I Would still call it `NodeSelector` instead of just `Selector` it better conveys the purpose.

fossedihelm replied:
thanks! If you agree, I opted for `NodeLabelSelector` :)

fossedihelm commented on cmd/virt-handler/virt-handler.go:
```diff
@@ -572,6 +578,44 @@ func (app *virtHandlerApp) shouldEnablePersistentReservation() {
+func (app *virtHandlerApp) shouldEnableKSM() {
+	currentKSMValue, err := os.ReadFile("/sys/kernel/mm/ksm/run")
+	if err != nil {
+		// Only enable for ksm-available nodes
```
Thanks, right, I will update it soon

fossedihelm replied:
In order to avoid continuous logging, I moved the read only when we really want to enable or disable ksm

fossedihelm commented on tests/infra_test.go:
```diff
@@ -1690,6 +1690,112 @@ var _ = Describe("[Serial][sig-compute]Infrastructure", Serial, decorators.SigCo
+					},
+				}
+			}),
+			Entry("the MatchExpressions selector of the ksmConfiguration", func() *v1.KSMConfiguration {
```
Added

fossedihelm commented on cmd/virt-handler/virt-handler.go:
```diff
@@ -735,3 +779,40 @@ func copy(sourceFile string, targetFile string) error {
 	return nil
 }
+
+func enableKSMAndPatchNode(virtClient kubecli.KubevirtClient, node *k8sv1.Node, currentValue []byte) {
```
Done

fossedihelm commented:
@acardace I addressed the comments. PTAL. Thanks

acardace commented on cmd/virt-handler/virt-handler.go:
```diff
@@ -572,6 +578,38 @@ func (app *virtHandlerApp) shouldEnablePersistentReservation() {
+func (app *virtHandlerApp) shouldEnableKSM() {
+	node, err := app.virtCli.CoreV1().Nodes().Get(context.Background(), app.HostOverride, metav1.GetOptions{})
+	if err != nil {
+		log.DefaultLogger().Errorf("An error occurred while retrieving the node list: %s", err)
```
Here the log message should be something like `An error occurred while retrieving the node object` since it's a `Get()` and not a `List()`?

acardace commented:
/approve

xpivarc reviewed (changes requested):
There is a possibility that as this is written a lot of `get` calls could be done at the same time ( O(n) - n=#nodes). Could we rather re-use some of the already available routines? (node-labeler maybe)?

fossedihelm commented:
> There is a possibility that as this is written a lot of `get` calls could be done at the same time ( O(n) - n=#nodes). Could we rather re-use some of the already available routines? (node-labeler maybe)?

I didn't realize that the node labeller immediately reacts to the config changes. I think it is better. Thank you!

xpivarc commented on automation/test.sh:
```diff
@@ -481,6 +481,19 @@ spec:
 EOF
 fi
 
+ksm_available_nodes_count=0
```
By misconfiguration I mean a case where I forgot to enable KSM in the `project-infra`. With this, you would skip the tests but that would be wrong because I would rather get feedback, "You forgot about KSM!". Therefore all I am asking is rather not have this and move the responsibility of skip to elsewhere.

fossedihelm replied:
Well, this is not related to enable ksm, but if it is available.
Are you suggesting to switch bask to `Skip()` inside the tests?

acardace replied:
I think KSM is always compiled in the kernel for CentOS, so not having it would be pretty hard, that said here probably anything works, even just always adding the KSM ginkgo label without detecting anything.

xpivarc replied:
>Are you suggesting to switch bask to Skip() inside the tests?

Not really. I am suggesting to not do any auto-detections. We should stick to what we do now, for example, add skip for lanes.
It is the same as you would be detecting if there is KVM or GPU... You never know if you did provision the right hardware or if the components that expose the resources just did not die. It is better to fail than a false skip.

fossedihelm replied:
Got it, thanks!

fossedihelm commented:
@xpivarc I moved the logic inside the node-labeller to save `get` api calls. PTAL
@acardace Since this is not what you approved anymore, I would ask you another review.
Thanks

acardace commented:
/remove-approve

xpivarc commented on pkg/virt-handler/node-labeller/node_labeller.go:
```diff
@@ -433,3 +434,70 @@ func (n *NodeLabeller) loadKSM() {
+// Empty Selector will enable ksm for every node
+func (n *NodeLabeller) shouldEnableKSM(node *v1.Node) {
+	ksmConfig := n.clusterConfig.GetKSMConfiguration()
+	if ksmConfig == nil {
```
Is this needed? I believe `LabelSelectorAsSelector` will return a `selector` which will match nothing and therefore line 454 should take care of this.

fossedihelm replied:
Without this line 448 will generate a nil pointer dereference error.
```
selector, err := metav1.LabelSelectorAsSelector(ksmConfig.NodeLabelSelector)
```

xpivarc commented on pkg/virt-handler/node-labeller/node_labeller.go:
```diff
@@ -433,3 +434,70 @@ func (n *NodeLabeller) loadKSM() {
+// shouldEnableKSM will update the ksm of the node (if available) based on the kv configuration.
+// If the node labels match the selector terms, the ksm will be enabled.
+// Empty Selector will enable ksm for every node
+func (n *NodeLabeller) shouldEnableKSM(node *v1.Node) {
```
Please add unit tests.

fossedihelm replied:
Done

xpivarc commented:
/approve
The approach looks good to me

fossedihelm commented:
/hold
I notice that current implementation requires two cycles for enable ksm and label the node where ksm is enabled.
This can be improved

fossedihelm commented:
@xpivarc @acardace  [Changes](https://github.com/kubevirt/kubevirt/compare/0a6c557507b0311ede92fe05bf0dbe58dbb8a99c..ad9330749cad1efe853fc5d45a9844f340215f2e)
- Moved the `loadKSM` function under the enable/disable ksm one (This allows to enable and add label in one shot).
- Switch using `n.KSM.SysfsFilePath` which was already there and allows us to create custom unit tests.
- Added unit tests
PTAL! Thanks

fossedihelm commented:
/unhold

fossedihelm commented:
@jean-edouard Can you have a look?

fossedihelm commented:
Rebased

jean-edouard commented on pkg/virt-handler/node-labeller/node_labeller.go:
```diff
@@ -433,3 +433,70 @@ func (n *NodeLabeller) loadKSM() {
+}
+
+func (n *NodeLabeller) enableKSM(node *v1.Node) {
+	currentKSMValue, err := os.ReadFile(n.KSM.SysfsFilePath)
```
Can't we run `loadKSM()` before `shouldEnableKSM()` and just check `n.KSM.Enabled` here?

fossedihelm replied:
Great catch! Thanks! But for this I also need to introduce `Available` field. I also arranged the function name. PTAL

jean-edouard commented on staging/src/kubevirt.io/api/core/v1/types.go:
```diff
@@ -2585,10 +2590,19 @@ type NodeMediatedDeviceTypesConfig struct {
 	MediatedDevicesTypes []string `json:"mediatedDevicesTypes,omitempty"`
 	// +optional
-	// +listType=atomic
+	//+listType=atomic
```
Probably better to not do that, all other comments do start with a space

fossedihelm replied:
Mmmmmh, this could be an outcome of a rebase. Thanks!

jean-edouard reviewed (commented):
Looks good, just a couple comments

jean-edouard reviewed (approved):
@acardace @xpivarc please ensure the latest version still makes sense
/hold

acardace commented:
LGTM, @xpivarc I'll leave it to you to unhold.

acardace commented:
/retest-required

fossedihelm commented:
/retest-required

fossedihelm commented:
Thank you everyone for your review
